### PR TITLE
Internal/add digital

### PR DESCRIPTION
**Pull request by texasaggie97**
- [X] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/nimi-python/blob/master/CONTRIBUTING.md).
- [X] I've updated [CHANGELOG.md](https://github.com/ni/nimi-python/blob/master/CHANGELOG.md) if applicable.
- [X] ~~I've added tests applicable for this pull request~~
### What does this Pull Request accomplish?
* Add very basic NI-DIGITAL support
### What testing has been done?
* Visual inspection of generated files
* Manual testing
* Flake8

### CODE REVIEW

File: build/helper/codegen_helper.py [hunk 1]
```diff
@@ -251,8 +253,8 @@ def _get_ctype_variable_definition_snippet_for_scalar(parameter, parameters, ivi
         S160. Input is size of input buffer:                                       visatype.ViInt32(0 if list is None else len(list))
         S170. Input is size of output buffer with mechanism ivi-dance, QUERY_SIZE: visatype.ViInt32()
         S180. Input is size of output buffer with mechanism ivi-dance, GET_DATA:   visatype.ViInt32(error_code)
-        S190. Input is size of output buffer with mechanism passed-in:             visatype.ViInt32(buffer_size)
-        S200. Output scalar or enum:                                               visatype.ViInt32()
+        S210. Input is size of output buffer with mechanism passed-in:             visatype.ViInt32(buffer_size)
```
Comment (marcoskirsch) [inline build/helper/codegen_helper.py]: Looks like S190 is still used but it's gone from documentation?
I really think that adding `ivi-dance-with-a-twist` support (with NI-FAKE coverage) should be its own PR separate from adding nidigital.

File: build/helper/codegen_helper.py [hunk 2]
```diff
@@ -293,15 +295,27 @@ def _get_ctype_variable_definition_snippet_for_scalar(parameter, parameters, ivi
                     definition = '{0}.{1}(error_code)  # case S180'.format(module_name, parameter['ctypes_type'])
                 else:
                     assert False, "ivi_dance_step {0} not valid for parameter {1} with ['size']['mechanism'] == 'ivi-dance'".format(ivi_dance_step, parameter['name'])
+            elif corresponding_buffer_parameters[0]['size']['mechanism'] == 'ivi-dance-with-a-twist':  # We are only looking at the first one. Assumes all are the same here, assert below if not
+                # Verify all corresponding_buffer_parameters are 'out' and 'ivi-dance'
```
Comment (marcoskirsch) [inline build/helper/codegen_helper.py]: did you mean 'ivi-dance-with-a-twist'?

File: build/helper/metadata_add_all.py [hunk 3]
```diff
@@ -337,6 +338,10 @@ def add_all_function_metadata(functions, config):
 
     for f in functions:
         _add_codegen_method(functions[f])
+        # NI-Digital doesn't have any documentation yet, so make sure the
```
Comment (marcoskirsch) [inline build/helper/metadata_add_all.py]: Is this meant to be permanent?
If so, I'd reword "Some drivers may not have documentation, so make sure the documentation key exists".
If not, then mark with TODO

File: build/helper/metadata_find.py [hunk 4]
```diff
@@ -5,17 +5,17 @@ def find_parameter(name, parameters):
     return parameter[0]
 
 
-def find_size_parameter(parameter_list, parameters):
+def find_size_parameter(parameter_list, parameters, key='value'):
     '''Returns the parameter that is used to specify the size other parameters. Applies to 'ivi-dance' and 'passed-in'.'''
```
Comment (marcoskirsch) [inline build/helper/metadata_find.py]: Does this apply to `ivi-dance-with-a-twist`? update comment.

Why is the key different in some cases? It'd be nice if it didn't need to be the case.

File: docs/_static/nidigital_usage.inc [hunk 5]
```diff
@@ -0,0 +1,15 @@
+Usage
+------
+
+The following is a basic example of using the **nidcpower** module to open a session to a Source Meter Unit and measure voltage and current.
```
Comment (marcoskirsch) [inline docs/_static/nidigital_usage.inc]: update comment

File: generated/nidmm/session.py [hunk 6]
```diff
@@ -1894,11 +1894,11 @@ def fetch_waveform(self, array_size, maximum_time=datetime.timedelta(millisecond
         '''
         vi_ctype = _visatype.ViSession(self._vi)  # case S110
         maximum_time_ctype = _converters.convert_timedelta_to_milliseconds(maximum_time, _visatype.ViInt32)  # case S140
-        array_size_ctype = _visatype.ViInt32(array_size)  # case S190
+        array_size_ctype = _visatype.ViInt32(array_size)  # case S210
         waveform_array_size = array_size  # case B600
         waveform_array_array = array.array("d", [0] * waveform_array_size)  # case B600
         waveform_array_ctype = get_ctypes_pointer_for_buffer(value=waveform_array_array, library_type=_visatype.ViReal64)  # case B600
-        actual_number_of_points_ctype = _visatype.ViInt32()  # case S200
+        actual_number_of_points_ctype = _visatype.ViInt32()  # case S220
```
Comment (marcoskirsch) [inline generated/nidmm/session.py]: was this renumbering necessary?
If it was, it'd be great to get that out of the way in a separate PR that does just that in preparation for this.

File: src/nidigital/examples/nidigital_self_cal_and_test.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/python
```
Comment (marcoskirsch) [inline src/nidigital/examples/nidigital_self_cal_and_test.py]: I'd like to review this in its own PR please.

File: src/nidigital/metadata/functions.py [hunk 8]
```diff
@@ -0,0 +1,4665 @@
+# -*- coding: utf-8 -*-
```
Comment (marcoskirsch) [inline src/nidigital/metadata/functions.py]: Let's add NI-provided (not the add-on) metadata to source control in its own PR.

File: src/nidigital/system_tests/test_system_nidigital.py [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+# import math
+import nidigital
+# import numpy
+# import pytest
+# import tempfile
+# import time
+
+
+def test_pass():
+    pass
```
Comment (marcoskirsch) [inline src/nidigital/system_tests/test_system_nidigital.py]: Add a TODO here please. Consider opening an issue that we need to do this.

File: generated/nidigital/session.py [hunk 10]
```diff
@@ -0,0 +1,4165 @@
...
+import nidigital.enums as enums
+import nidigital.errors as errors
+
+# Used for __repr__
+import pprint
+pp = pprint.PrettyPrinter(indent=4)
+
+
+# Helper functions for creating ctypes needed for calling into the driver DLL
+def get_ctypes_pointer_for_buffer(value=None, library_type=None, size=None):
```
Comment (epage) [inline generated/nidigital/session.py]: Is this and other related functions public?

File: src/nidigital/metadata/functions.py (hunk 8 above)
Comment (texasaggie97) [reply src/nidigital/metadata/functions.py]: Why? It's all needed.

File: src/nidigital/examples/nidigital_self_cal_and_test.py (hunk 7 above)
Comment (texasaggie97) [reply src/nidigital/examples/nidigital_self_cal_and_test.py]: Made the example a do_nothing example. We need _a_ file here that passes flake8 to make the build process happy.

File: src/nidigital/metadata/functions.py (hunk 8 above)
Comment (marcoskirsch) [reply src/nidigital/metadata/functions.py]: At the end of the day it's all needed. But if you post the NI-provided metadata in one PR, the reviewer doesn't have to look at all those file after every diff. Consuming the metadata is more likely to require multiple rounds, and it's very helpful to not have the trivial stuff in the way.

**Comment by marcoskirsch** [comment]
> Any example code to look at to get a feel for the API?

Not yet, at this point the goal is to codegen from metadata as-is, however it comes out and with no Python API-specific provisions.

File: src/nidigital/metadata/functions.py (hunk 8 above)
Comment (texasaggie97) [reply src/nidigital/metadata/functions.py]: Sure, why not have another PR...

File: CHANGELOG.md [hunk 11]
```diff
@@ -51,6 +51,13 @@ All notable changes to this project will be documented in this file.
     * #### Added
     * #### Changed
     * #### Removed
+* ### NI DIGITAL
```
Comment (marcoskirsch) [inline CHANGELOG.md]: Should be NI-Digital or NI-Digital Pattern Driver (second one more correct, but definitely missing the - )

File: Makefile [hunk 12]
```diff
@@ -1,5 +1,5 @@
 
-ALL_DRIVERS := nifake nidcpower nidmm nifgen niscope niswitch nise nimodinst
+ALL_DRIVERS := nifake nidigital nidcpower nidmm nifgen niscope niswitch nise nimodinst
```
Comment (marcoskirsch) [inline Makefile]: The order here is random.
I had sort of reordered alphabetically when I did merges but that has the hidden behavior of changing the overall version with whatever is last on the list... just an observation.

A comment, even attached to a new issue about "last one wins" could be a good idea but feel free to drop.

File: README.rst [hunk 13]
```diff
@@ -105,6 +105,47 @@ NI-DCPower Python API Status
 
 
 
+NI-Digital Pattern Driver Python API Status
+-------------------------------------------
+
++---------------------------------------+--------------------------+
+| NI-Digital Pattern Driver (nidigital) |                          |
```
Comment (marcoskirsch) [inline README.rst]: This should say "NI-Digital Pattern Based Driver"
